train.py

```diff
@@ -0,0 +1,249 @@
+    spacy_de: spacy.Language,
+    spacy_en: spacy.Language,
+    cfg: DictConfig,
+    device="cpu",
+) -> tuple[nn.Module, str]:
+    print(f"Train using {device}")
+    if device != "cpu":
+        torch.cuda.set_device(device)
+    run_id = initialize_wandb(cfg)
+
+    # Get the index for padding token
+    pad_idx = vocab_tgt["<blank>"]
+
+    # define model parameters and create the model
+    model = make_model(
+        len(vocab_src),
+        len(vocab_tgt),
+        n=cfg.model.n,
+        d_model=cfg.model.d_model,
+        d_ff=cfg.model.d_ff,
+        h=cfg.model.h,
+        dropout=cfg.model.dropout,
+    )
+    if device != "cpu":
+        model.cuda(device)
+    module = model
+
+    # Set LabelSmoothing as a criterion for loss calculation
+    criterion = LabelSmoothing(size=len(vocab_tgt), padding_idx=pad_idx, smoothing=0.1)
+    if device != "cpu":
+        criterion.cuda(device)
+    # Create dataloaders
+    train_dataloader, valid_dataloader = create_dataloaders(
+        vocab_src,
+        vocab_tgt,
+        spacy_de,
+        spacy_en,
+        device=device,
+        slice=cfg.data_slice,
+        batch_size=cfg.batch_size,
+        max_padding=cfg.max_padding,
+    )
+
+    # Define optimizer and learning rate scheduler
+    optimizer = torch.optim.Adam(model.parameters(), lr=cfg.base_lr, betas=(0.9, 0.98), eps=1e-9)
+    lr_scheduler = LambdaLR(
+        optimizer=optimizer,
+        lr_lambda=lambda step: rate(step, cfg.model.d_model, factor=1, warmup=cfg.warmup),
+    )
+
+    train_state = TrainState()
+
+    for epoch in range(cfg.num_epochs):
+        model.train()
+        print(f"Epoch {epoch}", flush=True)
+
+        # Train model for one epoch
+        t_loss, train_state = train_epoch(
+            train_dataloader,
+            model,
+            criterion,
+            optimizer,
+            lr_scheduler,
+            pad_idx=pad_idx,
+            train_state=train_state,
+            accum_iter=cfg["accum_iter"],
+        )
+
+        # Save checkpoint after each epoch
+        file_path = f"models/{cfg.file_prefix}-{run_id}-{epoch}.pt"
+        torch.save(
+            {
+                "model_state_dict": module.state_dict(),
+                "cfg": OmegaConf.to_container(cfg),
+            },
+            file_path,
+        )
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+
+        print(f"Epoch {epoch} Validation", flush=True)
+        model.eval()
+        # Evaluate the model on validation set
+        sloss = val_epoch(
+            train_dataloader,
+            model,
+            criterion,
+        )
+        # Log validation and training losses
+        print(sloss)
+        wandb.log({"val/loss": sloss, "train/loss": t_loss})
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+
+    return model, run_id
+
+
+def train_epoch(
+    data_iter: Iterable,
```

Comment:
```suggestion
    train_dataloader: Iterable,
```

If it's possible, it's better to keep the same name of a variable across multiple functions